src/models/tools/geometry/jxg-comment.ts

```diff
@@ -0,0 +1,137 @@
+import { JXGChangeAgent, JXGProperties } from "./jxg-changes";
+import { objectChangeAgent } from "./jxg-object";
+import { values } from "lodash";
+import { isMovableLine } from "./jxg-movable-line";
+import { isPolygon } from "./jxg-polygon";
+import { isBoard } from "./jxg-board";
+import { isPoint } from "./jxg-point";
+
+export const isCommentType = (v: any) => v && v.getAttribute("clientType") === "comment";
+
+export const isComment = (v: any) => isCommentType(v) && (v instanceof JXG.Text) && (v.elType === "text");
+
+export const getAnchor = (comment: JXG.Text) => {
+  if (!isComment) return;
+  const ancestors = values(comment.ancestors);
+  return ancestors.length === 1
+    ? ancestors[0]
+    : ancestors.find(elem => isMovableLine(elem) || isPolygon(elem));
+};
+
+const sharedProps = {
+  strokeWidth: 1,
+  clientType: "comment",
+  highlight: false,
+  strokeColor: "white",
+  clientStrokeColor: "white",
+  clientSelectedStrokeColor: "white",
+  fillColor: "red",
+};
+
+const pointProps = {
+  ...sharedProps,
+  withLabel: false,
+  visible: false
+};
+
+const lineProps = {
+  ...sharedProps,
+  strokeColor: "black",
+  strokeOpacity: .3,
+  straightFirst: false,
+  straightLast: false
+};
+
+function getCentroid(anchor: JXG.GeometryElement) {
+  if (isPoint(anchor)) {
+    const coords = (anchor as JXG.Point).coords.usrCoords;
+    return [coords[1], coords[2]];
+  } else if (anchor) {
+    const points = values(anchor.ancestors) as JXG.Point[];
+    const center = [0.0, 0.0];
+    points.forEach((point) => {
+      center[0] += point.coords.usrCoords[1];
+      center[1] += point.coords.usrCoords[2];
+    });
+    const len = points.length;
+    if (len) {
+      center[0] /= len;
+      center[1] /= len;
+    } else {
+      center[0] = NaN;
+      center[1] = NaN;
```

Comment:
Is NaN the most useful return here? I wonder if returning undefined rather than [NaN, NaN] might make more sense.

Reply:
Ah you're right, I copied the centroid code from that Google Group post but `undefined` is a better fit for us